### PR TITLE
Added DatabaseCredentials class for Postgres

### PR DESCRIPTION
**Pull request by crusher95**
Block used to manage authentication with a database.
    Attributes:
        database: The name of the database to use.
        username: The user name used to authenticate.
        password: The password used to authenticate.
        host: The host address of the database.
        port: The port to connect to the database.
        connect_args: The options which will be passed directly to the
            DBAPI's connect() method as additional keyword arguments.
    Example:
        Load stored database credentials:
        ```python
        from prefect_postgres import DatabaseCredentials
        database_block = DatabaseCredentials.load("BLOCK_NAME")
        ```

### CODE REVIEW

File: prefect_postgres/credentials.py [hunk 1]
```diff
@@ -1,3 +1,78 @@
...
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/3xLant5G70S4vJpmdWCYmr/8fdb19f15b97c3a07c3af3efde4d28fb/download.svg.png?h=250"  # noqa
+
+    username: Optional[str] = None
+    password: Optional[SecretStr] = None
+    database: Optional[str] = None
+    host: Optional[str] = None
+    port: Optional[str] = None
+    connect_args: Optional[Dict[str, Any]] = None
+
+    def get_connection(self) -> connect:
```
Comment (desertaxle) [inline prefect_postgres/credentials.py:40]: I think this needs to be typed as `AsyncConnection` class.
```suggestion
    def get_connection(self) -> psycopg.AsyncConnection:
```

File: prefect_postgres/credentials.py [hunk 2]
```diff
@@ -1,3 +1,78 @@
...
+                    cur.execute("
+                        CREATE TABLE test (
+                            id serial PRIMARY KEY,
+                            num integer,
+                            data text)
+                        ")
+            postgres_credentials_flow()
+            ```
+        """
+        return psycopg.connect(
```
Comment (desertaxle) [inline prefect_postgres/credentials.py:71]: We'll want to use `AsyncConnection` here to make sure that we can execute queries and operations asynchronously.
```suggestion
        return psycopg.AsyncConnection.connect(
```

File: prefect_postgres/credentials.py [hunk 3]
```diff
@@ -1,3 +1,78 @@
...
+    """
+    Block used to manage authentication with a database.
+    Attributes:
+        database: The name of the database to use.
+        username: The user name used to authenticate.
+        password: The password used to authenticate.
+        host: The host address of the database.
+        port: The port to connect to the database.
+        connect_args: The options which will be passed directly to the
+            DBAPI's connect() method as additional keyword arguments.
```
Comment (desertaxle) [inline prefect_postgres/credentials.py:21]:
```suggestion
           psycopg connect() method as additional keyword arguments.
```

**Review by desertaxle — CHANGES_REQUESTED** [review]
In addition to the specific comments left, can you please write some tests covering the new functionality and add an entry to CHANGELOG.md describing the changes?

File: prefect_postgres/credentials.py [hunk 4]
```diff
@@ -1,3 +1,73 @@
...
+
+    _block_type_name = "Postgres Database Credentials"
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/7G8c2Zz4j0yyhXqQ44SI2A/eee9ba482dd6b61862b588b6fd28ad81/PostgreSQL-Logo.png?h=250"  # noqa
+
+    username: Optional[str] = Field(default=None, description="The name of the database to use.")
+    password: Optional[SecretStr] = Field(default=None, description="The user name used to authenticate.")
+    database: Optional[str] = Field(default=None, description="The password used to authenticate.")
+    host: Optional[str] = Field(default=None, description="The host address of the database.")
+    port: Optional[str] = Field(default=None, description="The port to connect to the database.")
+    connect_args: Optional[Dict[str, Any]] = Field(default=None, title="Additional Configuration", description="Additional configuration to use when creating a database connection.")
```
Comment (zzstoatzz) [inline prefect_postgres/credentials.py:38]: Does it make sense for these to be `Optional`? `get_connection` appears to require all of them

Comment (desertaxle) [reply prefect_postgres/credentials.py:38]: Good call. We aren't accepting a connection URL so they are all required except for `connect_args` which should default to an empty dictionary.
```suggestion
    username: str = Field(default=None, description="The name of the database to use.")
    password: SecretStr = Field(default=None, description="The user name used to authenticate.")
    database: str = Field(default=..., description="The password used to authenticate.")
    host: str = Field(default=..., description="The host address of the database.")
    port: str = Field(default=..., description="The port to connect to the database.")
    connect_args: Dict[str, Any] = Field(default_factory=dict, title="Additional Configuration", description="Additional configuration to use when creating a database connection.")
```

File: .gitignore [hunk 5]
```diff
@@ -136,3 +136,116 @@ dmypy.json
...
+# See https://youtrack.jetbrains.com/issue/JBR-2257
+.idea/$CACHE_FILE$
+
+# CodeStream plugin
+# https://plugins.jetbrains.com/plugin/12206-codestream
+.idea/codestream.xml
+
+# Azure Toolkit for IntelliJ plugin
+# https://plugins.jetbrains.com/plugin/8053-azure-toolkit-for-intellij
+.idea/**/azureSettings.xml
```
Comment (ahuang11) [inline .gitignore:251]: If it was just one line, would this still work?
```suggestion
# PyCharm
.idea/
```

File: prefect_postgres/credentials.py [hunk 6]
```diff
@@ -1,3 +1,73 @@
...
+    Block used to manage authentication with a PostgreSQL database.
+    Attributes:
+        database: The name of the database to use.
+        username: The user name used to authenticate.
+        password: The password used to authenticate.
+        host: The host address of the database.
+        port: The port to connect to the database.
+        connect_args: The options which will be passed directly to the
+           psycopg connect() method as additional keyword arguments.
+    Example:
```
Comment (ahuang11) [inline prefect_postgres/credentials.py:22]:
```suggestion
    Block used to manage authentication with a PostgreSQL database.

    Attributes:
        database: The name of the database to use.
        username: The user name used to authenticate.
        password: The password used to authenticate.
        host: The host address of the database.
        port: The port to connect to the database.
        connect_args: The options which will be passed directly to the
           psycopg connect() method as additional keyword arguments.

    Example:
```

File: prefect_postgres/credentials.py [hunk 7]
```diff
@@ -1,3 +1,73 @@
 """
 Module containing functionality for authenticating with PostgreSQL databases
 """
+import psycopg
+from typing import Any, Dict, Optional, Union
+from pydantic import SecretStr
+from prefect.blocks.core import Block
+from psycopg import connect
```
Comment (ahuang11) [inline prefect_postgres/credentials.py:8]: Running `pre-commit install && pre-commit run --all` will run isort and sort this out and remove unused imports.

File: prefect_postgres/credentials.py [hunk 8]
```diff
@@ -1,3 +1,73 @@
 """
 Module containing functionality for authenticating with PostgreSQL databases
 """
+import psycopg
```
Comment (ahuang11) [inline prefect_postgres/credentials.py:4]:
```suggestion
from psycopg import AsyncConnection
```

File: prefect_postgres/credentials.py [hunk 9]
```diff
@@ -1,3 +1,73 @@
...
+                    host="localhost",
+                    port=5432
+                )
+                conn = postgres_credentials.get_connection()
+                with conn.cursor() as cur:
+                    cur.execute("CREATE TABLE test (id serial PRIMARY KEY,num integer,data text)")
+            postgres_credentials_flow()
+            ```
+        """
+        return psycopg.AsyncConnection.connect(
```
Comment (ahuang11) [inline prefect_postgres/credentials.py:66]:
```suggestion
        return AsyncConnection.connect(
```

File: prefect_postgres/credentials.py [hunk 10]
```diff
@@ -1,3 +1,73 @@
...
+                    password="prefect_password",
+                    database="postgres",
+                    host="localhost",
+                    port=5432
+                )
+                conn = postgres_credentials.get_connection()
+                with conn.cursor() as cur:
+                    cur.execute("CREATE TABLE test (id serial PRIMARY KEY,num integer,data text)")
+            postgres_credentials_flow()
+            ```
```
Comment (ahuang11) [inline prefect_postgres/credentials.py:64]:
```suggestion
        Returns an authenticated connection object that can be
        used to query from databases.

        Returns:
            The authenticated Psycopg3 AsyncConnection.

        Examples:
            Create an asynchronous connection to PostgreSQL using URL params.
            ```python
            from prefect import flow
            from prefect_postgres import DatabaseCredentials
            @flow
            def postgres_credentials_flow():
                postgres_credentials = DatabaseCredentials(
                    username="prefect",
                    password="prefect_password",
                    database="postgres",
                    host="localhost",
                    port=5432
                )
                conn = postgres_credentials.get_connection()
                with conn.cursor() as cur:
                    cur.execute("CREATE TABLE test (id serial PRIMARY KEY,num integer,data text)")
            postgres_credentials_flow()
            ```
```

File: prefect_postgres/credentials.py [hunk 11]
```diff
@@ -1,3 +1,73 @@
...
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/7G8c2Zz4j0yyhXqQ44SI2A/eee9ba482dd6b61862b588b6fd28ad81/PostgreSQL-Logo.png?h=250"  # noqa
+
+    username: str = Field(default=None, description="The name of the database to use.")
+    password: SecretStr = Field(default=None, description="The user name used to authenticate.")
+    database: str = Field(default=..., description="The password used to authenticate.")
+    host: str = Field(default=..., description="The host address of the database.")
+    port: str = Field(default=..., description="The port to connect to the database.")
+    connect_args: Dict[str, Any] = Field(default_factory=dict, title="Additional Configuration", description="Additional configuration to use when creating a database connection.")
+
+    def get_connection(self) -> psycopg.AsyncConnection:
```
Comment (ahuang11) [inline prefect_postgres/credentials.py:40]:
```suggestion
    def get_connection(self) -> AsyncConnection:
```

**Review by ahuang11 — CHANGES_REQUESTED** [review]
Looks like a good start. Some comments regarding styling and linting

**Comment by ahuang11** [comment]
I was just taking a look at https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.connect and if I'm not mistaken, I don't think you can directly pass all those keyword arguments directly into `AsyncConnection.connect`.

```python
>>> import asyncio
>>> from psycopg import AsyncConnection
>>> asyncio.run(AsyncConnection.connect(database="my_database"))
Traceback (most recent call last):
  File "/Users/andrew/mambaforge/envs/hightouch2/lib/python3.9/site-packages/psycopg/conninfo.py", line 96, in _parse_conninfo
    return pq.Conninfo.parse(conninfo.encode())
  File "/Users/andrew/mambaforge/envs/hightouch2/lib/python3.9/site-packages/psycopg/pq/pq_ctypes.py", line 927, in parse
    raise exc
psycopg.OperationalError: invalid connection option "database"


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/andrew/mambaforge/envs/hightouch2/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/andrew/mambaforge/envs/hightouch2/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/Users/andrew/mambaforge/envs/hightouch2/lib/python3.9/site-packages/psycopg/connection_async.py", line 122, in connect
    conninfo = make_conninfo(**params)
  File "/Users/andrew/mambaforge/envs/hightouch2/lib/python3.9/site-packages/psycopg/conninfo.py", line 59, in make_conninfo
    _parse_conninfo(conninfo)
  File "/Users/andrew/mambaforge/envs/hightouch2/lib/python3.9/site-packages/psycopg/conninfo.py", line 98, in _parse_conninfo
    raise e.ProgrammingError(str(ex))
psycopg.ProgrammingError: invalid connection option "database"
```

**Pull request closed without merging**